ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -162,52 +216,83 @@ densityDisconnect ::
   => GenesisWindow
   -> SecurityParam
   -> Map peer (AnchoredFragment (Header blk))
-  -> Map peer (Tip blk)
+  -> Set peer
   -> Map peer SlotNo
   -> AnchoredFragment (Header blk)
   -> ([peer], Map peer (DensityBounds blk))
-densityDisconnect (GenesisWindow sgen) (SecurityParam k) candidateSuffixes theirTips latestSlots loeFrag =
+densityDisconnect (GenesisWindow sgen) (SecurityParam k) candidateSuffixes caughtUpPeers latestSlots loeFrag =
   (losingPeers, densityBounds)
   where
     densityBounds = Map.fromList $ do
       (peer, fragment) <- Map.toList competingFrags
       -- Skip peers that haven't advertised their tip yet.
       -- They should be disconnected by timeouts instead.
-      theirTip <- toList (theirTips Map.!? peer)
+      -- TODO We probably need an alternative approach for this now
+      -- that we don't use the tip anymore
```

Comment:
If we adopt https://github.com/IntersectMBO/ouroboros-consensus/pull/984#discussion_r1521388834, we could check whether the `lastSlot` for the peer is `Origin`.

Reply:
That still wouldn't tell us if the peer is done at `Origin` or just uninitialized...

Reply:
hmm are you thinking of:
* no entry in `Map` -> uninitialized
* `Origin` -> tip was advertised and it is `Origin`
?

Reply:
FTR: `Maybe (WithOrigin SlotNo)` would be expressive enough to capture the three possibilities:

 - `Nothing`: Not initialized (`MsgIntersectFound` not yet received by the server). We should ignore such a peer in this GDD round.
 - `Just s`: Initialized, we can do density disconnections involving this peer.